elki-outlier/src/main/java/elki/outlier/isolation/IsolationForest.java

```diff
@@ -0,0 +1,366 @@
+
+  /**
+   * Class logger
+   */
+  private static final Logging LOG = Logging.getLogger(IsolationForest.class);
+
+  /*
+   * The set of trees
+   */
+  protected List<Tree> trees = null;
+
+  /*
+   * The number of trees
+   */
+  protected int numTrees = 100;
+
+  /*
+   * The sub sample size
+   */
+  protected int subsampleSize = 256;
+
+  /**
+   * Random generator
+   */
+  private RandomFactory rnd;
+
+  /**
+   * Constructor.
+   *
+   * @param numTrees
+   * @param subSamplingSize
+   * @param rnd
+   */
+  public IsolationForest(int numTrees, int subSamplingSize, RandomFactory rnd) {
+    this.numTrees = numTrees;
+    this.subsampleSize = subSamplingSize;
+    this.rnd = rnd;
+  }
+
+  public OutlierResult run(Database db, Relation<? extends NumberVector> relation) {
+    // Output data storage
+    WritableDoubleDataStore scores = DataStoreUtil.makeDoubleStorage(relation.getDBIDs(), DataStoreFactory.HINT_DB);
+    FiniteProgress prog = LOG.isVerbose() ? new FiniteProgress("IsolationForest progress", relation.size(), LOG) : null;
+    // Track minimum and maximum scores
+    DoubleMinMax minmax = new DoubleMinMax();
+
+    // Reduce sub sample size if data is too small
+    if(relation.size() < subsampleSize) {
+      subsampleSize = relation.size();
+    }
+
+    {
+      final Random random = rnd.getSingleThreadedRandom();
+      // Generate trees
+      trees = new ArrayList<>(numTrees);
+      for(int i = 0; i < numTrees; i++) {
+        ModifiableDBIDs sample = DBIDUtil.randomSample(relation.getDBIDs(), subsampleSize, rnd);
+        trees.add(new Tree( //
+            sample, //
+            random, //
+            0, //
+            (int) FastMath.ceil(FastMath.log(relation.size()) / FastMath.log(2)), //
+            relation));
+        LOG.incrementProcessed(prog);
+      }
+      LOG.ensureCompleted(prog);
+
+      // Iterate over all objects
+      for(DBIDIter iter = relation.iterDBIDs(); iter.valid(); iter.advance()) {
+        final double score = distributionForInstance(relation.get(iter));
+        scores.putDouble(iter, score);
+        minmax.put(score);
+        LOG.incrementProcessed(prog);
+      }
+      LOG.ensureCompleted(prog);
+    }
+
+    // Wrap the result in the standard containers
+    OutlierScoreMeta meta = new BasicOutlierScoreMeta(
+        // Actually observed minimum and maximum values
+        minmax.getMin(), minmax.getMax(),
+        // Theoretical minimum and maximum: no variance to infinite variance
+        0, Double.POSITIVE_INFINITY);
+    DoubleRelation rel = new MaterializedDoubleRelation("iforest anomaly score", relation.getDBIDs(), scores);
+    return new OutlierResult(meta, rel);
+  }
+
+  /**
+   * Returns the average path length of an unsuccessful search. Returns 0 if
+   * argument is less than or equal to 1.
+   */
+  protected static double c(double n) {
+    if(n <= 1.0) {
+      return 0;
+    }
+    return 2 * (FastMath.log(n - 1) + 0.5772156649) - (2 * (n - 1) / n);
```

Comment:
Undocumented "magic" constants such as `0.5772156649` are considered a bad style. Please replace this with a named constant from `MathUtil` (I don't think we have the Euler constant there yet).